File: eco-core/core-plugin/src/main/java/com/willfp/reforges/display/ReforgesDisplay.java
package com.willfp.reforges.display;

import com.willfp.eco.core.EcoPlugin;
import com.willfp.eco.core.display.Display;
import com.willfp.eco.core.display.DisplayModule;
import com.willfp.eco.core.display.DisplayPriority;
import com.willfp.eco.util.SkullUtils;
import com.willfp.reforges.reforges.Reforge;
import com.willfp.reforges.reforges.meta.ReforgeTarget;
import com.willfp.reforges.reforges.util.ReforgeUtils;
import org.apache.commons.lang.WordUtils;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.SkullMeta;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ReforgesDisplay extends DisplayModule {
    /**
     * Create weapons display.
     *
     * @param plugin Instance of Reforges.
     */
    public ReforgesDisplay(@NotNull final EcoPlugin plugin) {
        super(plugin, DisplayPriority.HIGHEST);
    }

    @Override
    protected void display(@NotNull final ItemStack itemStack,
                           @NotNull final Object... args) {
        ReforgeTarget target = ReforgeTarget.getForMaterial(itemStack.getType());

        if (target == null && itemStack.getType() != Material.PLAYER_HEAD) {
            return;
        }

        ItemMeta meta = itemStack.getItemMeta();
        assert meta != null;
        List<String> lore = meta.hasLore() ? meta.getLore() : new ArrayList<>();
        assert lore != null;

        Reforge reforge = ReforgeUtils.getReforge(meta);
        Reforge stone = ReforgeUtils.getReforgeStone(meta);

        if (reforge == null && stone == null) {
            if (this.getPlugin().getConfigYml().getBool("reforge.show-reforgable")) {
                List<String> addLore = new ArrayList<>();

                for (String string : this.getPlugin().getConfigYml().getStrings("reforge.reforgable-suffix")) {
                    addLore.add(Display.PREFIX + string);
                }

                lore.addAll(addLore);
            }
        }

        if (stone != null) {
            meta.setDisplayName(this.getPlugin().getConfigYml().getString("reforge.stone.name").replace("%reforge%", stone.getName()));
            SkullUtils.setSkullTexture((SkullMeta) meta, stone.getConfig().getString("stone-config.texture"));
            itemStack.setItemMeta(meta);
            List<String> stoneLore = new ArrayList<>();
            for (String string : this.getPlugin().getConfigYml().getStrings("reforge.stone.lore")) {
                stoneLore.add(Display.PREFIX + string.replace("%reforge%", stone.getName()));
            }
            lore.addAll(0, stoneLore);
        }

        if (reforge != null) {
            if (this.getPlugin().getConfigYml().getBool("reforge.display-in-lore")) {
                List<String> addLore = new ArrayList<>();

                addLore.add(" ");
                addLore.add(reforge.getName());

                List<String> description = new ArrayList<>(Arrays.asList(WordUtils.wrap(
                        reforge.getDescription(),
                        this.getPlugin().getConfigYml().getInt("reforge.line-wrap"),
                        "\n",
                        false
                ).split("\\r?\\n")));
                description.replaceAll(s -> this.getPlugin().getLangYml().getString("description-color") + s.replace("%description%", reforge.getDescription()));
                description.replaceAll(s -> s.replace("§r", this.getPlugin().getLangYml().getString("description-color")));
                addLore.addAll(description);

                addLore.replaceAll(s -> Display.PREFIX + s);
                lore.addAll(addLore);
            }
        }

        meta.setLore(lore);
        itemStack.setItemMeta(meta);
    }
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/reforges/display/ReforgesDisplay.kt
package com.willfp.reforges.display

import com.willfp.eco.core.display.Display
import com.willfp.eco.core.display.DisplayModule
import com.willfp.eco.core.display.DisplayPriority
import com.willfp.eco.core.display.DisplayProperties
import com.willfp.eco.core.fast.FastItemStack
import com.willfp.eco.core.fast.fast
import com.willfp.eco.core.placeholder.context.placeholderContext
import com.willfp.eco.util.SkullUtils
import com.willfp.eco.util.StringUtils
import com.willfp.eco.util.formatEco
import com.willfp.eco.util.toJSON
import com.willfp.libreforge.ItemProvidedHolder
import com.willfp.libreforge.ProvidedHolder
import com.willfp.reforges.ReforgesPlugin
import com.willfp.reforges.reforges.ReforgeTargets
import com.willfp.reforges.util.reforge
import com.willfp.reforges.util.reforgeStone
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import org.bukkit.persistence.PersistentDataType

@Suppress("DEPRECATION")
class ReforgesDisplay(private val plugin: ReforgesPlugin) : DisplayModule(plugin, DisplayPriority.HIGH) {
    private val tempKey = plugin.namespacedKeyFactory.create("temp")

    override fun display(
        itemStack: ItemStack,
        player: Player?,
        props: DisplayProperties,
        vararg args: Any
    ) {
        val targets = ReforgeTargets.getForItem(itemStack)

        val fast = itemStack.fast()

        val stone = fast.persistentDataContainer.reforgeStone

        if (targets.isEmpty() && stone == null) {
            return
        }

        val fastItemStack = FastItemStack.wrap(itemStack)

        val lore = fastItemStack.lore

        val reforge = fast.persistentDataContainer.reforge

        val context = placeholderContext(
            player = player,
            item = itemStack
        )

        if (reforge == null && stone == null) {
            if (plugin.configYml.getBool("reforge.show-reforgable")) {
                if (props.inGui) {
                    return
                }

                val addLore: MutableList<String> = ArrayList()
                for (string in plugin.configYml.getFormattedStrings("reforge.reforgable-suffix")) {
                    addLore.add(Display.PREFIX + string)
                }
                lore.addAll(addLore)
            }
        }

        if (stone != null) {
            val meta = itemStack.itemMeta
            meta.setDisplayName(stone.config.getFormattedString("stone.name"))
            val stoneMeta = stone.stone.itemMeta
            if (stoneMeta is SkullMeta) {
                val stoneTexture = SkullUtils.getSkullTexture(stoneMeta)

                if (stoneTexture != null) {
                    SkullUtils.setSkullTexture(meta as SkullMeta, stoneTexture)
                }
            }

            itemStack.itemMeta = meta

            val stoneLore = stone.config.getStrings("stone.lore")
                .map { it.replace("%price%", if (player == null) "" else stone.stonePrice?.getDisplay(player) ?: "") }
                .formatEco(player)
                .map { "${Display.PREFIX}$it" }

            lore.addAll(0, stoneLore)
        }

        if (reforge != null) {
            if (plugin.configYml.getBool("reforge.display-in-lore")) {
                val addLore: MutableList<String> = ArrayList()
                for (string in plugin.configYml.getFormattedStrings("reforge.reforged-prefix")) {
                    addLore.add(Display.PREFIX + string.replace("%reforge%", reforge.name))
                }
                addLore.addAll(reforge.description.formatEco(context))
                addLore.replaceAll { "${Display.PREFIX}$it" }
                lore.addAll(addLore)
            }

            if (plugin.configYml.getBool("reforge.display-in-name")) {
                val displayName = fastItemStack.displayNameComponent

                if (!fastItemStack.displayName.contains(reforge.name)) {
                    fastItemStack.persistentDataContainer.set(
                        tempKey,
                        PersistentDataType.STRING,
                        displayName.toJSON()
                    )

                    val newName = reforge.namePrefixComponent.append(displayName)

                    fastItemStack.setDisplayName(newName)
                }
            }


            if (player != null) {
                val provided = ItemProvidedHolder(reforge, itemStack)

                val lines = provided.getNotMetLines(player).map { Display.PREFIX + it }

                if (lines.isNotEmpty()) {
                    lore.add(Display.PREFIX)
                    lore.addAll(lines)
                }
            }
        }

        fastItemStack.lore = lore
    }

    override fun revert(itemStack: ItemStack) {
        itemStack.reforge ?: return

        val fis = FastItemStack.wrap(itemStack)

        if (!plugin.configYml.getBool("reforge.display-in-name")) {
            return
        }

        if (fis.persistentDataContainer.has(tempKey, PersistentDataType.STRING)) {
            fis.setDisplayName(
                StringUtils.jsonToComponent(
                    fis.persistentDataContainer.get(
                        tempKey,
                        PersistentDataType.STRING
                    )
                )
            )

            fis.persistentDataContainer.remove(tempKey)
        }
    }
}
